File: UnBeatables.cpp
#include <csignal>
#include <thread>

#include <Logs/EasyLogging.h>
#include <Logs/NaoqiLog.cpp>
#include <Config/INIParser.cpp>
#include "ConnectionConfig.hpp"
#include "perception/Perception.hpp"
#include "UnBoard.hpp"
#include "behavior/Behavior.hpp"
#include "communication/Communication.hpp"

INITIALIZE_EASYLOGGINGPP

PerceptionBoard perceptionBoard;
CommunicationBoard communicationBoard;


NAOqiAddress naoqiAddress;
GameControllerAddress gameControllerAddress;
qi::SessionPtr session;

std::shared_ptr<Perception> perception;
std::shared_ptr<Behavior> behavior;
std::shared_ptr<Communication> communication;

void close() {
    if (perception) {
        LOG(INFO) << "\x1B[32m[MAIN] Closing perception\x1B[0m";
        perception->close();
    }
    if (behavior) {
        LOG(INFO) << "\x1B[32m[MAIN] Closing behavior\x1B[0m";
        behavior->close();
    }
    if (session) {
        LOG(INFO) << "\x1B[32m[MAIN] Closing session\x1B[0m";
        session->close();
    }
}

void signalHandler(int signum) {
    LOG(WARNING) << "\x1B[32m[MAIN] Process interrupted by signal " << signum << "\x1B[0m";
    close();
    exit(signum);
}

void initLoggingSystem() {
    el::Loggers::reconfigureAllLoggers(el::ConfigurationType::Format,
    "\x1B[36m %datetime %level [%logger]\x1B[0m \t%msg");
    redirectNaoqiLogsToEasyLogging();
}

void initSession() {
    session = qi::makeSession();
    LOG(INFO) << "\x1B[32m[MAIN] Connecting to " << naoqiAddress.ip << ":" << naoqiAddress.port;
    auto connectFuture = session->connect("tcp://" + naoqiAddress.ip + ":" + std::to_string(naoqiAddress.port));

    if (connectFuture.wait(3000) == qi::FutureState_Running)
        throw std::runtime_error("Connection timeout");
    if (connectFuture.hasError())
        throw std::runtime_error(connectFuture.error());

    LOG(INFO) << "\x1B[32m[MAIN] Successfully connected to NAOqi\x1B[0m";
}

void initPerception() {
    perception = std::make_shared<Perception>();
    std::thread t(&Perception::process, perception);
    t.join();
}

void initBehavior() {
    behavior = std::make_shared<Behavior>();
    std::thread t(&Behavior::process, behavior);
    t.join();
}

void initCommunication() {
    communication = std::make_shared<Communication>();
    std::thread t(&Communication::process, communication);
    t.join();
}

int main() {
    std::signal(SIGINT, signalHandler);
    initLoggingSystem();
    loadConfig(naoqiAddress, gameControllerAddress);
    LOG(INFO) << "\x1B[32m[MAIN] Initializing Unboard\x1B[0m";

    try {
        // initSession();
        // initPerception();
        // initBehavior();
        initCommunication();
    }
    catch (const std::exception& ex) {
        LOG(ERROR) << "\x1B[31m[MAIN] Error while connecting or executing commands on NAO: " << ex.what() << "\x1B[0m";
        close();
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

File: UnBeatables.py
import signal
import sys
import threading
import logging
import qi

import UnBoard
from Libs.Ini.iniParser import ConfigLoader
from Libs.Logs import logs

from Behavior.behavior import Behavior
from Communication.communication import Communication
from Motion.motion import Motion
from Perception.perception import Perception

logging.basicConfig(level=logging.INFO)

class UnBeatables:
    def __init__(self, ip="127.0.0.1", port=9559):
        self.ip = ip
        self.port = port
        self.session = qi.Session()
        try:
            self.session.connect("tcp://{0}:{1}".format(self.ip, self.port))
            logs.logger_main.info("Connected to NAOqi at {0}:{1}".format(self.ip, self.port))
        except RuntimeError:
            logs.logger_main.exception("Could not connect to NAOqi at {0}:{1}".format(self.ip, self.port))
            sys.exit(1)

        self.motion = Motion(self.session)
        self.behavior = Behavior(self.session)
        self.perception = Perception(self.session)
        self.communication = Communication(self.session)

        self.behavior_thread = None
        self.perception_thread = None
        self.communication_thread = None

    def start_threads(self):
        logs.logger_main.info("Starting threads...")
        self.behavior_thread = threading.Thread(target=self.behavior.run)
        self.perception_thread = threading.Thread(target=self.perception.run)
        self.communication_thread = threading.Thread(target=self.communication.run)

        self.behavior_thread.start()
        self.perception_thread.start()
        self.communication_thread.start()

    def join_threads(self):
        if self.behavior_thread:
            self.behavior_thread.join()
        if self.perception_thread:
            self.perception_thread.join()
        if self.communication_thread:
            self.communication_thread.join()

    def shutdown(self, *_):
        logs.logger_main.info("Ctrl+C detected, shutting down modules...")
        try: self.behavior.close()
        except Exception as e: logs.logger_main.error("Error closing Behavior: {0}".format(e))
        try: self.perception.close()
        except Exception as e: logs.logger_main.error("Error closing Perception: {0}".format(e))
        try: self.motion.close()
        except Exception as e: logs.logger_main.error("Error closing Motion: {0}".format(e))
        try: self.communication.close()
        except Exception as e: logs.logger_main.error("Error closing Communication: {0}".format(e))
        logs.logger_main.info("All modules closed. Exiting program.")
        sys.exit(0)


if __name__ == "__main__":
    loader = ConfigLoader("config.ini")
    loader.apply()

    nao = UnBeatables(ip=UnBoard.naoqiAddress_ip,
                      port=UnBoard.naoqiAddress_port)
    signal.signal(signal.SIGINT, nao.shutdown)
    nao.start_threads()
    signal.pause()
